pkg/services/ngalert/store/instance_database.go

```diff
@@ -149,12 +162,34 @@ func (st DBstore) FetchOrgIds(ctx context.Context) ([]int64, error) {
 	return orgIds, err
 }
 
-func (st DBstore) DeleteAlertInstance(ctx context.Context, orgID int64, ruleUID, labelsHash string) error {
-	return st.SQLStore.WithTransactionalDbSession(ctx, func(sess *sqlstore.DBSession) error {
-		_, err := sess.Exec("DELETE FROM alert_instance WHERE rule_org_id = ? AND rule_uid = ? AND labels_hash = ?", orgID, ruleUID, labelsHash)
+// DeleteAlertInstances deletes instances with the provided keys. It
+// writes transactions with a maximum count of 400 rows. Larger writes are
+// split into multiple transactions.
+func (st DBstore) DeleteAlertInstances(ctx context.Context, keys ...models.AlertInstanceKey) error {
```

Comment:
My latest change makes these both single transactions with many statements, instead of big concatenation statements. Benchmarking the two didn’t show much performance difference, and it lets us use a single transaction like we ought to.